### PR TITLE
JS Query: Missing index adjustment after concurrent modification

### PR DESCRIPTION
**Pull request by asger-semmle**
Flags array iteration code that removes an element using `splice`, but then accidentally skips the next element due to shifting.

Preliminary results from an older query on LGTM.com [look promising](https://lgtm.com/query/2436582718718897713/).

See [this example](https://lgtm.com/projects/g/tinymce/tinymce/snapshot/4f15a28ce42052b9cf8531425c12c38becd53c42/files/src/plugins/contextmenu/main/ts/ui/ContextMenu.ts?sort=name&dir=ASC&mode=heatmap#L34) and [this example](https://lgtm.com/projects/g/rapidpro/rapidpro/snapshot/7a6d1950aba0003c53f3a14ff1f5dcb2b07f4a3b/files/static/bower/bootstrap/docs/build/node_modules/hogan.js/lib/compiler.js?sort=name&dir=ASC&mode=heatmap#L76) for the motivation behind some of the whitelisting rules. The former of these is technically buggy if there are only two elements in the array, but that's out of scope for this query. The advice about decrementing the loop index would certainly not work as intended in that case

[This example](https://lgtm.com/projects/g/galaxyproject/galaxy/snapshot/f40b7210b5c230cb354007e57207a5487bb6fe76/files/client/galaxy/scripts/libs/jquery/select2.js?sort=name&dir=ASC&mode=heatmap#L3231) is interesting. They only decrement when `i > 0` for some reason, but I think we're right to flag it unless there is something special about the first element.

Evaluation underway.

### CODE REVIEW

**Comment by asger-semmle** [comment]
[Evaluation results](https://git.semmle.com/asger/dist-compare-reports/tree/Asgers-MacBook-Pro.local_1545088143310). I spotted one false positive in [three.js](https://github.com/bkimminich/juice-shop/blob/eebb83ccbaf1d03201c9e43b5df9bacc501a1be2/app/private/three.js#L30818) due to the chain of `break`s that depend on a variable being set in the inner loop. I can't see an obvious way to fix that.

[Performance](https://git.semmle.com/gist/asger/99b6987b6d6ae451b9e626db1feae44f) warrants a closer look.

File: javascript/ql/src/Statements/MissingIndexAdjustmentAfterConcurrentModification.qhelp [hunk 1]
```diff
@@ -0,0 +1,69 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+<p>
+  Items can be removed from an array using the <code>splice</code> method, but when doing so,
+  all subseequent items will be shifted to a lower index. If this is done while iterating over
```
Comment (xiemaisi) [inline javascript/ql/src/Statements/MissingIndexAdjustmentAfterConcurrentModification.qhelp]: "subsequent"

File: javascript/ql/src/Statements/MissingIndexAdjustmentAfterConcurrentModification.ql [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+ * @kind problem
+ * @problem.severity warning
+ * @id js/missing-index-adjustment-after-concurrent-modification
+ * @tags correctness
+ * @precision high
+ */
+import javascript
+
+/**
+ * Operation that inserts or removes elements from an array while shifting all elements
```
Comment (xiemaisi) [inline javascript/ql/src/Statements/MissingIndexAdjustmentAfterConcurrentModification.ql]: "An operation"

File: javascript/ql/src/Statements/MissingIndexAdjustmentAfterConcurrentModification.ql [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    or
+    hasPathThrough(splice, cfg.getAPredecessor()) and
+    getLoopEntry().dominates(cfg.getBasicBlock()) and
+    not hasIndexingManipulation(cfg)
+  }
+}
+
+from ArrayIterationLoop loop, SpliceCall splice
+where loop.hasPathThrough(splice, loop.getUpdate().getFirstControlFlowNode())
+select splice, "Missing loop index adjustment after removing array item. Some array items will be skipped due to shifting."
```
Comment (xiemaisi) [inline javascript/ql/src/Statements/MissingIndexAdjustmentAfterConcurrentModification.ql]: Could we make this alert message shorter, ideally just one sentence? Something like "Removing an array item without adjusting the loop index may cause subsequent items to be skipped" might do (and you could even incorporate the name of the loop variable into the message).

**Comment by asger-semmle** [comment]
I've address the comment, though it made me wonder if the query name is also too long and not to be point. Do you think something like "Loop iteration skipped due to shifting" would be better?

**Comment by xiemaisi** [comment]
>  Do you think something like "Loop iteration skipped due to shifting" would be better?

Yes, I like that.

File: javascript/config/suites/javascript/correctness-more [hunk 4]
```diff
@@ -12,6 +12,7 @@
 + semmlecode-javascript-queries/LanguageFeatures/InconsistentNew.ql: /Correctness/Language Features
 + semmlecode-javascript-queries/LanguageFeatures/SpuriousArguments.ql: /Correctness/Language Features
 + semmlecode-javascript-queries/Statements/MisleadingIndentationAfterControlStmt.ql: /Correctness/Statements
++ semmlecode-javascript-queries/Statements/MissingIndexAdjustmentAfterConcurrentModification.ql: /Correctness/Statements
```
Comment (xiemaisi) [inline javascript/config/suites/javascript/correctness-more]: Query name needs to be changed here as well.

File: javascript/ql/src/Statements/LoopIterationSkippedDueToShifting.qhelp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+</overview>
+<recommendation>
+
+<p>
+Determine what the loop is supposed to do:
+</p>
+
+<ul>
+  <li>
+    If the intention is to remove <em>every occurence</em> of a certain value, decrement the loop counter after removing an element, to counterbalance
```
Comment (mc-semmle) [inline javascript/ql/src/Statements/LoopIterationSkippedDueToShifting.qhelp]: occurrence (2 Cs, 2 Rs)

**Comment by asger-semmle** [comment]
@mc-semmle thanks for the review 👍

**Pull request merged**